src/coreclr/jit/fginline.cpp

```diff
@@ -586,8 +586,70 @@ class SubstitutePlaceholdersAndDevirtualizeWalker : public GenTreeVisitor<Substi
                 }
 
                 CORINFO_CONTEXT_HANDLE contextInput = context;
+                context                             = nullptr;
                 m_compiler->impDevirtualizeCall(call, nullptr, &method, &methodFlags, &contextInput, &context,
                                                 isLateDevirtualization, explicitTailCall);
+                if (context != nullptr && !m_compiler->compCurBB->HasFlag(BBF_INTERNAL))
+                {
+                    CORINFO_METHOD_INFO methInfo;
+                    if (m_compiler->info.compCompHnd->getMethodInfo(call->gtCallMethHnd, &methInfo, context) &&
+                        genActualType(JITtype2varType(methInfo.args.retType)) == genActualType(call->TypeGet()))
+                    {
+                        IL_OFFSET ilOffset = 0;
+                        INDEBUG(ilOffset = call->gtRawILOffset);
+                        CORINFO_CALL_INFO callInfo = {};
+                        callInfo.hMethod           = method;
+                        callInfo.methodFlags       = methodFlags;
+                        m_compiler->impMarkInlineCandidate(call, context, false, &callInfo, ilOffset);
+
+                        const bool isInlineCandidate                  = call->IsInlineCandidate();
+                        const bool isGuardedDevirtualizationCandidate = call->IsGuardedDevirtualizationCandidate();
```

Comment:
I don't see any reason to look for GDV cases here -- we're not going to rerun the GDV transforms (nor could we, without some considerable refactoring) and if the original call was a GDV we would have devirtualized it already.